### ISSUE TITLE
Provide line number and column in case of validation errors

### ISSUE THREAD

**Issue by mhoran** [yannh/kubeconform#156]
We are trying to use `kubeconform` as a library to build a [language server](https://langserver.org/) that will expose manifest errors in the IDE. In order to best surface these errors, the line number and column of validation errors needs to be available. Currently, `kubeconform` surfaces the path to the invalid node, but there are some issues with how this is presented. In addition, this currently requires parsing the error returned from the error object, rather than through a rich interface exposing the underlying details.

Currently the path to an invalid node is exposed as `foo.bar.baz`. However, this has some issues. This form is ambiguous, as the path corresponds to both of the following:
```
foo:
  bar.baz: "qux"
foo.bar:
  baz: "quux"
```

With `kubeval` we were able to change the path separator from `.` to `|` and leverage the `yaml.v3` to find the line number based on this path.

If `kubeconform` implemented its own `Error` and returned this with additional information, one that we could cast from an err.Error), then it would remain backwards compatible but allow downtream tools to extract extra information. It would be incredibly helpful (also to command line users) to expose the line and column number (which could be implemented by parsing into a `Node` using `yaml.v3`, and following the `JsonContext`), but even just exposing the `JsonContext` itself, to allow downstream tools to figure out things like the line/column number would be very helpful. In general, it would be great if some of the additional details that are jammed into the plain text error message were also available on this implementation to avoid needing to extract details from the error via regex.

We just wanted to get your thoughts on the best approach here given there are some alternative approaches (returning the line number and column from `kubeconform` itself, extending the error object, and/or returning the `JsonContext` via a custom error object.) There may be other approaches as well -- so it'd be good to align on what makes most sense before moving forward.

**Comment by hadar-co** [comment]
Hi, over at [Datree](https://datree.io) the CLI provides the line and column numbers of discovered misconfigurations in YAML files, using  `yaml.v3` and `yqlib v4`.

Have a look here:
https://github.com/datreeio/datree/blob/main/pkg/evaluation/evaluator.go#L418

**Comment by mhoran** [comment]
Thanks @hadar-co, that looks very similar to what we were doing with `kubeval` before experimenting with `kubeconform`. However, I believe the code in `datree` will be subject to the ambiguity issue presented in my original post.

As we're building a language server, we would also prefer to use a Go library -- which `kubeconform` supports ["experimentally"](https://github.com/yannh/kubeconform#using-kubeconform-as-a-go-module).

The simplest solution would be to extend the `Result` returned in `ValidateResource` here: https://github.com/yannh/kubeconform/blob/752a33eaeb225830fbf32b900331fc37c596d467/pkg/validator/validator.go#L203 with the errors returned by `Validate` at https://github.com/yannh/kubeconform/blob/752a33eaeb225830fbf32b900331fc37c596d467/pkg/validator/validator.go#L195

But given others likely would like access to the line number and character, perhaps the path parsing that both you and we are doing to find that information ourselves would be best handled in `kubeconform`.

In addition, `datree` and our code has to parse plain text error messages to extract information. However, it'd be nice if a "rich" `Error` interface were presented, to avoid the need for pulling out these details with a regexp. The same interface extension would be valuable for `yaml.v3`, since the line number of invalid YAML must be extracted from the error message itself.

**Comment by yannh** [comment]
This sounds interesting. If anyone wants to have a go at it, I'd definitely review. Otherwise I might get to it but no guarantees when :)

**Comment by yannh** [comment]
@mhoran kubeconform uses "gojsonschema" for the validation, and unfortunately it's not looking good, see this bug: https://github.com/xeipuuv/gojsonschema/issues/185

**Comment by mhoran** [comment]
Interesting. I suppose I understand the argument, and the proposed alternative does allow a consumer to highlight a line / block of text if appropriate. It's unfortunate though since this can only be done with an additional parser pass.

`kubeconform` could still do one of two things. Either pass the `JsonContext` back to the library consumer so that lines can be looked up by the caller, or do this lookup itself with an additional parser pass. The latter is what we've done ourselves in the language server with `kubeval`, however the path returned from `jsoncontext` by default is ambiguous and `kubeconform` does not allow changing the path separator as `kubeval` does.

I may have some time in the coming weeks to put a PR together. If there's a preferred direction (looking up line numbers in `kubeconform` vs simply passing the context back to the caller), let me know. Presumably line numbers would be useful to most `kubeconform`, but this may be a bit more work and could have a performance impact.

**Comment by yannh** [comment]
" however the path returned from jsoncontext by default is ambiguous and kubeconform does not allow changing the path separator as kubeval does"  can you give some more information on this?

Generally I'd be quite interested in having thie capability in kubeconform. If the performance impact is mostly when an error occurs, I'd be ok with that - assuming that in most workflows the number of valid files will outnumber the number of invalid files significantly.

**Comment by mhoran** [comment]
> " however the path returned from jsoncontext by default is ambiguous and kubeconform does not allow changing the path separator as kubeval does" can you give some more information on this?

Here's an example `ConfigMap` that shows the ambiguity:
```
apiVersion: v1
kind: ConfigMap
data:
  foo:  
    bar:
      baz
data.foo:
  bar   
```

```
[mhoran@ilectra] ~/t% kubeconform -strict test.yaml
test.yaml - ConfigMap  is invalid: For field (root): Additional property data.foo is not allowed - For field data.foo: Invalid type. Expected: [string,null], given: object
```
Because the default separator for `JsonContext.String()` is `.`, the path is ambiguous when a key actually contains a `.`. That could happen e.g. with annotations or elsewhere. This is a contrived example, but hopefully it helps convey the issue.

We worked around this with `kubeval` as the entire `ResultError` is returned on validation failure: https://github.com/instrumenta/kubeval/blob/fe0a7c22b93b92adfdc57d07b92d5231fd0b3e0e/kubeval/kubeval.go#L190

This `ResultError` allows [access](https://github.com/xeipuuv/gojsonschema/blob/b076d39a02e5015af0a2a96636e4cc479ecd9f45/result.go#L76) to the `JsonContext` where the separator can be [changed](https://github.com/xeipuuv/gojsonschema/blob/b076d39a02e5015af0a2a96636e4cc479ecd9f45/jsonContext.go#L43).

**Comment by yannh** [comment]
Gotcha - should we start there then? Bubble up the gojsonschema error? One thing I'd like somehow would be to keep backward compatibility of the "library" interface, as I know a few people rely on it.

**Comment by tomasaschan** [comment]
As long as whatever is returned from the validator implements `error` it should be source backwards compatible, right?

**Comment by mhoran** [comment]
That was my thinking. See https://go.dev/blog/module-compatibility#working-with-interfaces.

**Comment by yannh** [comment]
FYI I am doing some discovery to see how hard it would be to use https://github.com/santhosh-tekuri/jsonschema which seems better maintained. This might touch how this is implemented.

**Comment by yannh** [comment]
Here is the WIP https://github.com/yannh/kubeconform/pull/168/files , using this library which seems actively maintained, and at least on par performance and feature-wise:  https://github.com/santhosh-tekuri/jsonschema  
It mentions "rich, intuitive hierarchial error messages with json-pointers to exact location"  which I thought could be pretty useful to you.

**Comment by mhoran** [comment]
Looks good to me! I just tested it out and the path returned is a proper JSON pointer. The path is unambiguous in all cases, and a slash in the path is encoded as `~1`. Thanks for looking into this!

**Comment by yannh** [comment]
Note: some files might contain multiple resources, so not super sure how to get to a line number from a json schema :thinking:

**Comment by mhoran** [comment]
At least for our case we were planning to split the multi-resource documents ourselves; that was the only reliable way to get YAML syntax errors to be meaningful as well. With the new jsonschema library I get the following from kubeconform based when run against this document:
```
apiVersion: v1
kind: ConfigMap
data:
  foo:
    bar:
      baz
data.foo:
  bar
---
apiVersion: v1
kind: ConfigMap
data:
  foo:
    bar:
      baz
data.foo:
  bar

```
```
[mhoran@ilectra] ~/kubeconform% go run cmd/kubeconform/main.go -strict test.yaml
test.yaml - ConfigMap  is invalid: problem validating schema. Check JSON formatting: jsonschema: '/data/foo' does not validate with https://raw.githubusercontent.com/yannh/kubernetes-json-schema/master/master-standalone-strict/configmap-v1.json#/properties/data/additionalProperties/type: expected string or null, but got object
test.yaml - ConfigMap  is invalid: problem validating schema. Check JSON formatting: jsonschema: '/data/foo' does not validate with https://raw.githubusercontent.com/yannh/kubernetes-json-schema/master/master-standalone-strict/configmap-v1.json#/properties/data/additionalProperties/type: expected string or null, but got object
exit status 1
```

**Comment by yannh** [comment]
So, you would need the JSON path returned somehow by the "Validate" function, so you don't have to parse the output directly? I guess it would be nice to also expose this in the output, when running with output json?

**Comment by mhoran** [comment]
The changes in the PR are already a huge improvement, since the path is not ambiguous and I can parse it out. The previous library required access to the JsonContext in order to address the ambiguity so that's not an issue now.

Iif there were a rich error object (just implementing the Error interface for backwards compatibility) then I wouldn't even have to parse the error message to get the path. That'd be great -- but not strictly necessary at this point.

**Comment by yannh** [comment]
I'd rather avoid having people parse the output, since it's not something I can guarantee to be stable over time. If this unblocks you that's fine, but given how apparently datree also suggested the change in output breaks some stuff on their side (see the closed PR) you might not be the only one interested in that path :) I'll investigate, this is definitely sth I see to be in scope of this project :+1:

**Comment by yannh** [comment]
Released in v0.6.0 :rocket:

**Comment by levenleven** [comment]
@yannh Thanks for adding this! I've just tried to play with new version and it exposes the location in schema instead of path in validated document. Opened https://github.com/yannh/kubeconform/pull/177 to expose path in doc instead. Or maybe we should expose both?
